pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+		// from the dnsNames map. Also, send the DNS name to the deleted channel
+		// indicating that the details corresponding to the DNS name have been
+		// deleted. Otherwise, add the updated resolved name to the dnsNames map
+		// corresponding to the DNS name.
+		if !resolvedName.regularObjExists && !resolvedName.wildcardObjExists {
+			delete(res.dnsNames, dnsName)
+			res.deleted <- dnsName
+		} else {
+			res.dnsNames[dnsName] = resolvedName
+		}
+	}
+}
+
+// getNextDNSNameDetails returns the DNS name with minimum next lookup time.
+// It also returns the next lookup time and the number of IP addresses
+// associated with the DNS name. If no such DNS name exists, then false is
+// returned, otherwise true is returned.
+func (res *Resolver) getNextDNSNameDetails() (string, time.Time, int, bool) {
+	res.dnsLock.Lock()
+	defer res.dnsLock.Unlock()
+
+	exists := false
+	var minNextLookupTime time.Time
+	var dns string
+	var numIPs int
+
+	for dnsName, resolvedName := range res.dnsNames {
+		if !exists || resolvedName.minNextLookupTime.Before(minNextLookupTime) {
+			exists = true
+			minNextLookupTime = resolvedName.minNextLookupTime
+			dns = dnsName
+			numIPs = resolvedName.numIPs
+		}
+	}
+	return dns, minNextLookupTime, numIPs, exists
+}
+
+// lookupDNSName sends a DNS lookup request to CoreDNS pod(s). The DNS lookup is performed to
+// trigger an update, if required, of the DNSNameResolver resources matching the DNS name.
+func (res *Resolver) lookupDNSName(dnsName string, numIPs int) error {
+	// By default, the DNS lookup request will be sent to maxCoreDNSPodIPs number of CoreDNS
+	// pods.
+	numCoreDNSPodIPs := maxCoreDNSPodIPs
+	// If the DNS name has 0 or 1 associated IP addresses, then the DNS lookup request will
+	// be sent to only 1 CoreDNS pod.
+	if numIPs <= 1 {
+		numCoreDNSPodIPs = 1
+	}
+
+	// Get the randomly chosen CoreDNS pod IPs.
+	coreDNSPodIPs, err := res.getRandomCoreDNSPodIP(numCoreDNSPodIPs)
+	if err != nil {
+		return err
+	}
+
+	// Send the DNS lookup request to the CoreDNS pods for both A and AAAA type DNS records.
+	for _, recordType := range []uint16{dns.TypeA, dns.TypeAAAA} {
+		for _, coreDNSPodIP := range coreDNSPodIPs {
+			dnsMsg := &dns.Msg{}
+			dnsMsg.SetQuestion(dns.Fqdn(dnsName), recordType)
+			serverStr := net.JoinHostPort(coreDNSPodIP, defaultPort)
+			dnsClient := &dns.Client{
+				Timeout: defaultMinTTL,
+			}
+			if _, _, err := dnsClient.Exchange(dnsMsg, serverStr); err != nil {
+				logrus.Warningf("Failed to lookup DNS name: %s from CoreDNS pod with IP: %s, err: %s", dnsName, coreDNSPodIP, err)
+			}
+		}
+	}
+	return nil
+}
+
+// getRandomCoreDNSPodIP returns randomly chosen CoreDNS pod IPs. The input maxIPs defines
+// the upper limit for the number of CoreDNS pod IPs returned.
+func (res *Resolver) getRandomCoreDNSPodIP(maxIPs int) ([]string, error) {
+	// List all the CoreDNS pod endpointslices.
+	epList := &discoveryv1.EndpointSliceList{}
```

Comment:
No `dnsLock` here.

Reply:
Same as https://github.com/openshift/cluster-dns-operator/pull/394#discussion_r1459688287

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2